### PR TITLE
ISSUE-46-OCR: Provide the ability to OCR image files

### PR DESCRIPTION
**Pull request by patdunlavey**
Spitballing some code for OCR'ing image files:
 1. In OcrPostProcessor::buildExecutableCommand, we do a bunch of refactoring to enable it to generate tesseract commands when the input file is a raster type, while preserving the ability to use ghostscript to first rasterize pdf files.
 2. In OcrPostProcessor::settingsForm, we add a field for defining a file metadata field that contains the sequence number for the file 
3. In OcrPostProcessor::defaultConfiguration, we add 'configured_input_argument' => 'sequence_key'. This feeds into a change in AbstractPostProcessorQueueWorker.
4. In AbstractPostProcessorQueueWorker::invokeProcessor, we use the configured_input_argument if defined to grab the sequence number 
5. In StrawberryRunnersPostProcessorPluginBase, we add 'configured_input_argument' => '' to the default configuration.

@DiegoPino this code is not tested. Just wanting to get your reactions to approaches I'm playing with.

### CODE REVIEW

File: src/EventSubscriber/StrawberryRunnersEventSavePostProcessingSubscriber.php [hunk 1]
```diff
@@ -251,7 +251,7 @@ public function onEntitySave(StrawberryfieldCrudEvent $event) {
                     $valid_ado_type = array_map('trim', $valid_ado_type);
                     if (empty($config['ado_type']) || count(array_intersect($valid_ado_type, $sbf_type)) > 0) {
                       $valid_mimes = explode(',', $config['mime_type']);
-                      $valid_mimes = array_map('trim', $valid_mimes);
+                      $valid_mimes = array_filter(array_map('trim', $valid_mimes));
```
Comment (DiegoPino) [inline src/EventSubscriber/StrawberryRunnersEventSavePostProcessingSubscriber.php:254]: Good!

File: src/Plugin/StrawberryRunnersPostProcessor/OcrPostProcessor.php [hunk 2]
```diff
@@ -31,13 +31,16 @@
...
+  public $pluginDefinition;
+
   /**
    * {@inheritdoc}
    */
   public function defaultConfiguration() {
     return [
         'source_type' => 'asstructure',
         'mime_type' => ['application/pdf'],
+        'configured_input_argument' => 'sequence_key',
```
Comment (DiegoPino) [inline src/Plugin/StrawberryRunnersPostProcessor/OcrPostProcessor.php:43]: well well good!

**Comment by patdunlavey** [comment]
Of course!

**Comment by DiegoPino** [comment]
@patdunlavey I made a series of Changes but once I finally got your code working I'm having doubts about the sequence id approach.
- First, the data is coming from the file metadata and not the full JSON which means really we will either always have a sequence key at that level, so we really do not need a config for that. And the processor has access to that metadata... so its more like "take it from there" if not set ...
- the sequence id we us in the Solr Document ID is absolute to the same file. So an image file will always have 1 for that, since the file won't have multiple components.
- We can still save the relative sequence_id in the Solr document as a fallback/sort for a generic view
- But we need also in the Solr document either the "parent Object" sequence_id (for relative to a parent ordering)
- We might need to totally change the OCR itself to use instead of the Page Number (that is a big architectural change) here:
https://github.com/esmero/strawberry_runners/blob/f6cfcf783b55cc0a206d9436799acf81116fbb46/src/Plugin/StrawberryRunnersPostProcessor/OcrPostProcessor.php#L553 

Because basically the XML can not have sequence_and a number only... for highlighting i will really need to know the FILE uuid and a sequence_id ONLY for PDF generated content (when the sequence is not variable/metadata driven).

Only way I "could" get away with using the current XML page id is IF all my Solr queries are specific to a File ID ...and not general like "get me all for this NODE where processor is OCR"

@alliomeria @giancarlobi since you were kindly following with ideas and possible solutions

**Comment by giancarlobi** [comment]
@DiegoPino already read in slack thread and commented. Just a note on code line above, we need something like that also with ALTO, if I remember well. Sorry, my fault, to not be so present these days, I'll try to follow this as much as can.

**Comment by DiegoPino** [comment]
@giancarlobi thanks friend. Mostly I want you to be aware of the discussion and how this evolving

**Comment by patdunlavey** [comment]
@DiegoPino I've been asked to try to get this task going again. (It's HARD to come back to something that has been out of mind for a month!) It seems we got [side-tracked into the question of getting pdftoalto compiled for ARM64](https://archipelagocommons.slack.com/archives/C01A4MPFQA0/p1649093318019649), and that's where things stopped. So we never came back to this comment thread.

I wonder if you can take another stab advice on what I can do to move this forward?

**Comment by DiegoPino** [comment]
@patdunlavey been updating on slack about the pdfalto thing (really important) and its done. The main issue now is the language part which involved me refactoring your pull, but its Friday and I might have to wait until Monday (Maybe sooner) to pull against yours

**Comment by DiegoPino** [comment]
@patdunlavey and sorry for the delay, we really have our hands full, as I know you do to

**Comment by patdunlavey** [comment]
@DiegoPino I changed the target to 0.4.0. 

The only part of my code I'm aware of that I don't remember you providing review/feedback on is the update hook. I noticed that I had failed to include `datafolder_tesseract` in the settings, so that's fixed. I'm not sure if you'll like the way I found the path to the pdfalto executable. Can you take a look at that? If good then I hope we can move on with the merge!

**Comment by DiegoPino** [comment]
@patdunlavey checked (will also test  the end of the day -on a quite complex debug session and brain is full or objects that I do not understand-) and pretty sure all is well friend! Much better than how we started. If something is missing/lacking I follow up with a pull shortly testing live tonight, no worries. Thanks so much 🥳

**Review by DiegoPino — COMMENTED** [review]
Thanks for all your work on this!

**Pull request merged**